### PR TITLE
Automated tests package

### PR DESCRIPTION
**Pull request by mmarciniak90**
PR contains a package with automated tests.
Tests are connected with regression test cases.

#### Forms for testing
* [Espresso_forms.zip](https://github.com/opendatakit/collect/files/3177332/Espresso_forms.zip)

### CODE REVIEW

**Comment by lognaturel** [comment]
This is looking really good to me. I think it would be great to establish some conventions around where form XML, tests, helper methods, etc should go. Let's get on a call with anyone interested in testing after 1.22 is released.

**Comment by seadowg** [comment]
@lognaturel I'd be interested in joining for that!

File: collect_app/src/androidTest/java/org/odk/collect/android/BaseFormTest.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import static org.hamcrest.CoreMatchers.not;
+
+public class BaseFormTest {
+
+    @Rule
+    public ActivityTestRule<MainMenuActivity> main = new ActivityTestRule<>(MainMenuActivity.class);
+    @Rule
+    public GrantPermissionRule permissionRule = GrantPermissionRule.grant(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE);
+
+    protected static Matcher<View> withIndex(final Matcher<View> matcher, final int index) {
```
Comment (lognaturel) [inline collect_app/src/androidTest/java/org/odk/collect/android/BaseFormTest.java]: This is now available in `org.odk.collect.android.test`.

**Comment by lognaturel** [comment]
Call notes and action items [in this Google Doc](https://docs.google.com/document/d/1zVAieTlPoowK1Ab092AS1DHDAafQhuMJgmMdyE-hZRk/edit).

**Comment by seadowg** [comment]
[Here's](https://github.com/seadowg/taflan.) the example project (a boring CSV editor app) I was talking about.

The Espresso tests for the project are [here](https://github.com/seadowg/taflan/tree/master/src/androidTest/kotlin/com/seadowg/taflan/test). Looks like the pattern I used was roughly one "feature" of the app per class and then maybe more than one test in the class if there were "edge cases" or multiple paths to the feature. Sorry it's in Kotlin not Java!

Jonathan Gerrish from Google recently did a talk about building testable apps that includes a few examples of how they think about testing at different levels. Might not be the most relevant but you can find it [here](https://www.youtube.com/watch?v=VJi2vmaQe6w&list=PLOU2XLYxmsILVTiOlMJdo7RQS55jYhsMi&index=140&t=0s) if you're interested. You'll notice the video shows the new `ActivityScenario` I mentioned. This doesn't seem to be used in any of the official Espresso docs yet so maybe it's worth ignoring that at the moment.

**Comment by mmarciniak90** [comment]
@seadowg thanks for the shared information!
@lognaturel I updated PR.

We worked on refactoring our tests. If you have time, please take a look.

We were not able to copy form to SD card for now, but it is our next goal. We also want to better organize the `EspressoTestUtilities` class.

**Comment by seadowg** [comment]
@mmarciniak90 liking the first step of extracting those helpers to a new class. 

It might be nice to organize them into several classes - one for each "page" ("Form entry page" etc) so you only import the helpers for the pages your test cares about. That would also read quite nicely as you could end up with code that's like:

```java
MainMenu.startBlankForm("All widgets");
FormEntry.clickGoToIconInForm()
```

File: collect_app/src/androidTest/java/org/odk/collect/android/regression/EspressoTestUtilities.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
+package org.odk.collect.android.regression;
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/regression/EspressoTestUtilities.java]: I think this class should be one level up. You want to have all regression cases in one package what's ok but we will add more tests (out of this package) which could use the same helpers.

File: collect_app/src/androidTest/java/org/odk/collect/android/regression/BaseFormTest.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import android.Manifest;
+
+import androidx.test.rule.ActivityTestRule;
+import androidx.test.rule.GrantPermissionRule;
+
+import org.junit.Rule;
+import org.odk.collect.android.activities.MainMenuActivity;
+
+public class BaseFormTest {
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/regression/BaseFormTest.java]: If this class is used for all regression tests, not only those related just to form filling then we need a more generic name like BaseRegressionTest.

Comment (grzesiek2010) [reply collect_app/src/androidTest/java/org/odk/collect/android/regression/BaseFormTest.java]: and it should be abstract.

**Comment by grzesiek2010** [comment]
@mmarciniak90 @kkrawczyk123 I fixed the problem with copying forms to sd card, now you can use it in the same way you tried before (and like I did in my commit) and it will work.

**Comment by seadowg** [comment]
@grzesiek2010 @mmarciniak90 what was the problem around copying the forms? We found that there were problems with the `@BeforeClass` as it runs before any `@Rule`s (which we're using to give our self write permissions permission). 

Using the `CopyFormRule` like in #3161 might let you solve your issues without having to add code to store the form in the database?

**Comment by grzesiek2010** [comment]
The problem was that apart from copying files we need to save forms in the database to make them visible in the form list, it's here: https://github.com/opendatakit/collect/pull/3066/commits/c41252464dd20fe3f5956ace5f59977d4941a5ee#diff-95bba6de31f4296ffadad5868939acb3R77

If we download a form from a server we do both (copying forms and saving them to the database) so we need both here as well.
Of course, we have a task that would scan forms and add them to the database for us if we open the list of forms but waiting for such a task is problematic in tests, so doing the same we do while downloading forms is a better approach.

**Comment by seadowg** [comment]
@grzesiek2010 ah interesting. I'm guessing you're looking at the `DiskSyncTask` that the `FormChooserList` kicks off? 

What's weird is Espresso **should** wait for the `AsyncTask` thread pool to settle before running any matchers/actions (https://developer.android.com/training/testing/espresso/#sync) so [this line](https://github.com/opendatakit/collect/pull/3066/files#diff-c7d23d6b3c1d8251ff8726507c418a0dR50) should execute until the task has finished. Was that the line that was having problems?

**Comment by grzesiek2010** [comment]
Yes, it's that line. Maybe there is another fix but I think this is better because as I said if you download forms they are saved to the database so why we would need to test `DiskSyncTask` instead?

**Comment by seadowg** [comment]
@grzesiek2010 Yeah I hear what you're saying. I guess what I see are downsides to manually storing the form in the database (although not end of the worlds) is:

1. Our test now has to "fake" work that the app should be responsible for. Given we're thinking about these tests as "black box" then it would nice for us to avoid that kind of intervention as much as possible. It just introduces more of a chance that the test only passes because of the setup work we've done.
1. On the other side of the coin there is a risk that the test ends up breaking because we change something about how forms are saved. Again, not the end of the world, but it can be frustrating when you realize a test is failing because it is mimicking some functionality you've just changed in the app.
1. There is also a question around why we need to fake the form saving. It worries me a little that our app/tests aren't behaving as intended (i.e. syncing on AsyncTasks). That or maybe we don't understand Espresso properly.

What might be a nice step forward if we can't get it to work with Espresso without manually saving the form would be to share code between the `DiskSyncTask` and the test as much as possible (i.e. both call a `syncFormsOnDisk` method). That way anything that changes will be reflected in both places.

**Comment by grzesiek2010** [comment]
`DiskSyncTask` is used only if you put new forms in `odk/forms/` folder manually, in all cases where forms are downloaded from a server or form GoogleDrive they are saved in a similar way I used (not using  `DiskSyncTask`), that's why I think it's more natural.

The code responsible for copying files and saving data to the database is not big even if we change it it won't be a big deal to fix tests.

**Comment by mmarciniak90** [comment]
@grzesiek2010 thanks for help with forms. I still see a problem when attachments are required in the form. Could you look at this? 

@seadowg thanks for the advice with split the EspressoTestUtilities class. I introduced these changes.

**Comment by seadowg** [comment]
Could we rebase this on top of master? I was having a play with the form copying problems and it would be good to pull in the new changes to `FormLoadingUtils`.

**Comment by lognaturel** [comment]
Good work going on here, thank you!

I was originally imagining that tests that involve filling out a form could be organized around the form they fill. Similarly to the existing tests in the `formentry` package, they could each handle putting their form on the sdcard and letting the `DiskSyncTask` take care of loading it. As @grzesiek2010 mentions, this is the code path used when forms are pushed manually to a device. Launching `FormEntryActivity` with an intent extra that specifies which form to load is the code path used by external applications that use Collect for form filling.

I now see that conceptual test cases may span more than one form so that approach won't meet the requirements. There does need to be a way to explicitly trigger forms being put in the database from these tests. I agree with @seadowg that ideally this would use real application code. Could that involve unifying `DiskSyncTask.buildContentValues` and `FormDownloader.saveNewForm` and then putting that in a place that the tests can also use?

How about rebasing this on master, merging it and then proposing a way to do that cleanup in another PR?

**Comment by grzesiek2010** [comment]
> thanks for help with forms. I still see a problem when attachments are required in the form. Could you look at this?

Fixed.

> How about rebasing this on master

Done.

**Comment by lognaturel** [comment]
Remaining todos I see are:
- [x] move helpers up one level so all tests can use them
- [x] consider renaming the base test and making it abstract
- [ ] doing some refactoring so application code can be used to populate the form definition database

Does that sound right? If so, let's give @mmarciniak90 a chance to take a look at the first two when she's available, merge, and then follow up with another PR for the third.

**Comment by mmarciniak90** [comment]
@grzesiek2010 I noticed unexpected behavior. Could you look at it if you have time?
1. I used a fresh app (I reset whole settings). 
2. I run test `CascadingSelectWithNumberInHeaderTest` where I needed to use this code which copies form file and attachment.

`    public RuleChain copyFormChain = RuleChain`
            `.outerRule(GrantPermissionRule.grant(`
                   ` Manifest.permission.READ_EXTERNAL_STORAGE,`
                   ` Manifest.permission.WRITE_EXTERNAL_STORAGE,`
                 `   Manifest.permission.READ_PHONE_STATE)`
          `  )`
           ` .around(new CopyFormRule("numberInCSV.xml", "regression",`
`Collections.singletonList("itemSets.csv")));`

3. Test passed but when I looked into ODK files I noticed extra folders and files with prefix 'forms'



Additionally, I'm not sure why permission 'READ_PHONE_STATE' is needed.

**Comment by grzesiek2010** [comment]
Ok, I'll look into it.

**Comment by seadowg** [comment]
@grzesiek2010 @mmarciniak90 it looks like that is probably caused by line 114 in `FormLoadingUtils`:

```
String mediaPathName = Collect.FORMS_PATH + formFilename.replace(".xml", "") + FileUtils.MEDIA_SUFFIX + "/";
```

There should be a `File.seperator` after the `Collect.FORMS_PATH`.

**Comment by lognaturel** [comment]
> There should be a File.seperator after the Collect.FORMS_PATH.

Ahh, yes. Consider cherry-picking https://github.com/opendatakit/collect/pull/3175/commits/8743a8c0fbbf2654c7faa8379747e77ebf4356a6 if you want or doing something similar in your own commit and I can deal with merge conflicts once this is in. Even more reason to get this merged soon!

**Comment by mmarciniak90** [comment]
> move helpers up one level so all tests can use them

Moved to `collect_app/src/androidTest/java/org/odk/collect/android/espressoutils/`

> consider renaming the base test and making it abstract

Renamed to BaseRegressionTest
Consulted with @grzesiek2010 and we decided to leave it as a not abstract class

**Review by grzesiek2010 — APPROVED** [review]

**Comment by grzesiek2010** [comment]
Ok, I just added one commit which improved the way we clear the app state after tests and it's ready I think so merged!
